### PR TITLE
Allow overriding  and disabling the death formspec.

### PR DESCRIPTION
**Pull request by red-001**

### CODE REVIEW

**Comment by orwell96** [comment]
Wouldn't it be better to handle the death formspec entirely on server side?
- player dies
- server sends 'dead' packet (hp 0)
- client freezes movement/digging/interacting and tints screen red, but does nothing else
- server sends formspec
- client clicks a button in formspec
- server sends 'respawn' packet (hp 20) and a setpos
- client is happy and disables the red tint and movement/action freeze

**Comment by red-001** [comment]
@orwell96 Damage form falling and lava is handled by the client so it wouldn't make much sense to rewrite just the death formspec without also fixing that.

**Comment by est31** [comment]
I'd like if this is more compatible with #4737 so that in the protocol an empty formspec means that its disabled. Basically remove the "enable" flag while keeping the use_custom flag.

**Comment by Fixer-007** [comment]
Will this help fix this? -> https://github.com/minetest/minetest_game/issues/1216

**Comment by orwell96** [comment]
@est31 but then remember to not open the form when there's an empty string, the changes I did will not handle this.

**Comment by rubenwardy** [comment]
Feature freeze is now in effect

**Comment by bigfoot547** [comment]
I made a mod for this, it uses @orwell96 's `close_formspec` method, and use that in the `on_die` callback.

File: src/network/clientpackethandler.cpp [hunk 1]
```diff
@@ -846,6 +846,18 @@ void Client::handleCommand_InventoryFormSpec(NetworkPacket* pkt)
...
 }
 
+void Client::handleCommand_DeathScreenFormSpec(NetworkPacket* pkt)
+{
+	LocalPlayer *player = m_env.getLocalPlayer();
+	assert(player != NULL);
+
+	// Store formspec in LocalPlayer
+	*pkt >> player->custom_deathscreen;
+	*pkt >> player->enable_deathscreen;
```
Comment (nerzhul) [inline src/network/clientpackethandler.cpp]: please invert the two fields, enable and then the string

File: src/player.h [hunk 2]
```diff
@@ -166,6 +166,11 @@ class Player
 
 	std::string inventory_formspec;
 
+	bool custom_deathscreen;
+	bool enable_deathscreen;
+	std::string get_deathscreen_formspec() { return deathscreen_formspec; }
```
Comment (nerzhul) [inline src/player.h]: this function should be const. Also please attach deathscreen to SAO not Player we are trying to reduce/remove the Player class server side

Comment (nerzhul) [inline src/player.h]: also return a const ref to prevent copying data, this permit to enhance a little bit the performance

File: src/player.h [hunk 3]
```diff
@@ -166,6 +166,11 @@ class Player
...
+	void set_deathscreen_formspec(std::string input) { deathscreen_formspec = input; }
```
Comment (nerzhul) [inline src/player.h]: we don't pass std::string as copy if they are not modified before doing things. Please use const std::string & instead

File: src/player.h [hunk 4]
```diff
@@ -188,6 +193,7 @@ class Player
 	// hud for example can be modified by EmergeThread
 	// and ServerThread
 	Mutex m_mutex;
+	std::string deathscreen_formspec;
```
Comment (nerzhul) [inline src/player.h]: This variable is not initialized in constructor

**Comment by red-001** [comment]
closing this as this is something that should be done using client-sided modding

**Pull request closed without merging**